[jit] support torch.as_tensor in script

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23266 [jit] Support nn.GRU in script
* #23249 [jit] Support pack_padded_sequence and pad_packed_sequence
* #23248 [jit] support torch._C._get_tracing_state in script
* **#23247 [jit] support torch.as_tensor in script**

Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

wanchaol committed:
[jit] support torch.as_tensor in script

eellison reviewed (commented):
I haven't really looked into it but i think all of the implementation is more or less a copy-pasta of the torch.tensor code. Can you refactor the code ?

wanchaol committed:
refactor on "[jit] support torch.as_tensor in script"

[jit] support torch.as_tensor in script

gh-metadata: pytorch pytorch 23247 gh/wanchaol/32/head

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -527,7 +527,11 @@ class ShapePropagator {
-      case aten::tensor: {
+      case aten::tensor:
+      case aten::as_tensor: {
+        if (node->inputs().at(0)->type()->isSubtypeOf(TensorType::get())) {
```
What is the point of grouping these two cases together and then always breaking in the case of `as_tensor` ?

eellison replied:
Maybe add that as_tensor has an overloaded schema and can either have a tensor or a list as the first input

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1544,6 +1548,27 @@ class ShapePropagator {
+          at::ScalarType default_type = type->scalarType();
+          c10::Device default_device = type->device();
+          if (auto dtype_index = node->schema().argumentIndexWithName("dtype")) {
+            auto inp = toIValue(node->inputs().at(*dtype_index));
```
toIValue may return `c10::nullopt`, you still have to check for this case and add corresponding tests?

eellison replied:
if it's `c10::nullopt` you can't infer the output type here, so you have to return before setting the output type

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1544,6 +1548,27 @@ class ShapePropagator {
+          }
+          if (auto device_index = node->schema().argumentIndexWithName("device")) {
+              auto inp = toIValue(node->inputs().at(*device_index));
+            if (!inp->isNone()) {
```
same here for `c10::nullopt`

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -150,6 +150,68 @@ void recursiveStore(
+Operation createTensorFromList(const Node* node) {
+  auto input = node->inputs().at(0);
+  auto elem_type = input->type();
+  while (auto list_type = elem_type->cast<ListType>()) {
```
This doesn't work for as_tensor because `checkListInputType` expects a non-tensor input.

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -150,6 +150,68 @@ void recursiveStore(
+    IValue dtype;
+    IValue device;
+    if (if_set_requires_grad) {
+      pop(stack, data, dtype, device, requires_grad);
```
it is kind of hard to follow that `requires_grad` is only conditionally set if `if_set_requires_grad` a templated parameter is set to true.

Also In order to check if this function is correct you have to look back at all of its callsites and see if if the schema matches to verify correctness, which kind of breaks abstraction of the function.

wanchaol replied:
what do you mean by this? this abstraction only differed in a flag where one schema contains requires_grad and the other not

eellison replied:
maybe add comment that torch::tensor has a fourth `requires_grad` arg that `as_tensor` does not have

eellison reviewed (commented):
Looks good! i think there are a couple errors in shape analysis that need fixing.

Also in the interest of supporting `GRU` in script it's worth mentioning that you could rewrite:

`torch.as_tensor(lengths, dtype=torch.int64)` to `lengths.to(dtype=torch.int64)` which is currently supported

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -150,6 +150,68 @@ void recursiveStore(
+  while (auto list_type = elem_type->cast<ListType>()) {
+    elem_type = list_type->getElementType();
+  }
+  checkListInputType(elem_type, node);
```
`checkListInputType` will still output `torch.tensor` even if it's a call from `as_tensor`

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -326,60 +411,30 @@ RegisterOperators reg({
         "aten::tensor(t[] data, *, ScalarType? dtype=None, Device? device=None, bool requires_grad=False) -> Tensor",
+        createTensorFromList<true>),
+    Operator(
+        "aten::as_tensor(Tensor data, *, ScalarType? dtype=None, Device? device=None) -> Tensor",
```
The alias annotation here should be `Tensor(a) data) -> Tensor(a|b)`

wanchaol committed:
Update on "[jit] support torch.as_tensor in script"

wanchaol committed:
Update on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

wanchaol committed:
Update on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -283,6 +347,29 @@ RegisterOperators reg({
+      }),                                                                     \
+  Operator(                                                                   \
+      "aten::as_tensor(" #operator_type                                       \
+      " t, *, ScalarType? dtype=None, Device? device=None) -> Tensor",        \
```
Can you refactor this function with the equivalent torch::tensor one ?

wanchaol replied:
I was trying to do that, then I realized that there're macro's arg that could not be placed just as a template arg, like `tensor_creation_op`

wanchaol replied:
let me try it again to see if it works.

wanchaol replied:
Again failed because `tensor_creation_op` is a function call instance that could not make into a template arg.

eellison replied:
Yea makes sense. Could you factor out 
```
    at::ScalarType scalar_type =
        dtype.isNone() ? tensor.scalar_type() : dtype.toScalarType();
    c10::Device dev =
        device.isNone() ? tensor.device() : device.toDevice();
    if (scalar_type != initial_scalar_type || dev != tensor.device()) {
      tensor = tensor.to(dev, scalar_type);
    }
``` 
to a method ? those lines are repeated 3 times in the file.

eellison reviewed (commented):
Looks good, just one function that is copy-pasta'd to be refactored and should be ready to go

wanchaol committed:
Update on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

eellison reviewed (approved):
LGTM. One small comment about refactoring but should be good to go

wanchaol committed:
refactor on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -48,6 +48,20 @@ void checkListInputType(const c10::TypePtr& elem_type, const Node* node) {
+at::Tensor castTensorTo(at::Tensor self,
+                        const IValue& dtype,
+                        const IValue& device,
+                        at::ScalarType initial_scalar_type) {
```
i think you can just use can just use the `self.scalar_type()` don't need `initial_scalar_type`...

wanchaol replied:
Yeah I think it should be the same, but seems like `torch.tensor` also use initial_scalar_type that is getting from `scalarTypeFromJitType`, did you remember why we have this?

eellison replied:
eh doesn't really matter. the tensor was constructed from `initial_scalar_type` so it and `self.scalar_type()` should be same value

wanchaol committed:
Update on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)

wanchaol committed:
rebase with master on "[jit] support torch.as_tensor in script"


Differential Revision: [D16466590](https://our.internmc.facebook.com/intern/diff/D16466590)
